### PR TITLE
Add Mummer4 tools

### PR DESCRIPTION
**Pull request by nickeener**
Hey my name's Nick and this is my first project for Dr. Nekrutenko. It's not quite done and there are still a few problems that I will list, but Nekrutenko suggested making a pull request so more people could offer help.

This is my attempt at wrapping the 16 commands of the mummer4 package, which contains sequence alignment and visualization tools. Of the 16 commands, I was able to able to get 10 of them to pass the test run by planemo (nucmer, delta-filter, exact-tandems, mgaps, mummer, repeat-match, show-aligns, show-coords, show-diff, show-snps). Another 3 fail the test but that problem should be fixed by the following pull request by Dave B:  https://github.com/mummer4/mummer/pull/68 (promer, dnadiff, mummerplot).

Problems:
mummer - needs to incorporate data manager created by Dave B (https://github.com/davebx/tools-iuc/tree/mummer_dm/data_managers/data_manager_mummer4_index_builder). Not sure how to go about this.

dnadiff - First test gives error: "Syntax error: Unbound variable input". Not sure what's going on here. Second test fails due to foundation.pm error that's fixed by aforementioned pull request.

show-tiling - When running the command, the output is blank. I have have the dependencies and optional utilities installed so I don't know why it's not working.

combineMUMs - When I run this command I get a seg fault. Also no idea why.

annotate - Confused about the input for this command. Man page says it's the output from the gaps command except there is no gaps command and the mgaps output doesn't seem to work.

I'd appreciate any suggestions on fixing these problems or any concerning any of the other commands. I look forward to working with you all.

Nick Keener

### CODE REVIEW

**Comment by peterjc** [comment]
For reference, I wrapped a sub-set of the suite here:
http://toolshed.g2.bx.psu.edu/view/peterjc/mummer
https://github.com/peterjc/pico_galaxy/tree/master/tools/mummer

What you are doing here will be far more flexible and powerful - good plan 👍

**Comment by bgruening** [comment]
Wonderful @nickeener! Many of us are currently at the GCC ... so sorry if we take a little bit longer to review it. Thanks!

File: tools/mummer4/dnadiff.xml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            <param name="input_type" type="select" label="Choose input type">
+                <option value="sequence">Sequence data</option>
+                <option value="delta">Delta file from nucmer</option>
+            </param>
+            <when value="sequence">
+                <param name="reference_sequence" type="data" format="fasta" label="Reference Sequence" />
+                <param name="query_sequence" type="data" format="fasta" multiple="True" label="Query Sequence(s)" />
+            </when>
+            <when value="delta">
+                <param name="delta_data" type="data" format="txt" label="Alignment file from nucmer" />
```
[the remark this answers is not in the document]
Comment (nickeener) [reply tools/mummer4/dnadiff.xml]: How would I tell if they were tabular? The original file in this case had the suffix ".delta" which wasn't an accepted format so I changed it to .txt.

File: tools/mummer4/combineMUMs.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    	<param name="filename" type="text" argument="-W" value="witherrors.gaps" label="Output filename" />
+    	<param name="cover" type="boolean" argument="-x" truevalue="-x" falsevalue="" label="Don't output .cover files" />
+    	<param name="cutoff" type="float" argument="-e" min="0.0" max="100.0" value="0.02" label="Error-rate cutoff" />
+    </inputs>
+    <outputs>
+    	<data name="output" format="txt" from_work_dir="combineMUMs.txt" />
+    </outputs>
+    <tests>
+    	<test>
+    		<param name="reference_sequence" ftype="fasta" value="K12.fasta" />
```
[the remark this answers is not in the document]
Comment (nickeener) [reply tools/mummer4/combineMUMs.xml]: The indentation is fine in the file on my system so why does it appear wrong here?

**Comment by nickeener** [comment]
I have a weird problem with the mgaps command. I initially installed mummer4 on my laptop several months ago and mgaps worked fine. But then just recently I got a new laptop and installed mummer4 on it and now the mgaps command does not exist. Not sure why it didn't get installed the 2nd time or how to get it now.

Also, the nucmer command is now failing even though it passed before making this pull request and I have not updated it since.

**Comment by bgruening** [comment]
How did you installed it? With `conda create -n mummer mummer4`?

**Comment by nickeener** [comment]
I used: conda install -c bioconda mummer4
I have all the other commands, just not mgaps.

**Comment by nekrut** [comment]
Great job! Some notes:

 - [ ] Let's restrict the toolkit to `mummer`, `mummerplot`, `nucmer`, `nucplot`, `filter-delta`, `show-coord`
 - [ ] Combine `nucmer` and `nucplot` into one tool with an additional option to output the plot. Same thing for mummer

**Comment by nickeener** [comment]
@bgruening Could you please take a look at my modified code? Travis-CI tests do not pass yet due to some problem with git but we are working on that.

File: tools/mummer4/delta-filter.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+<tool id="delta-filter" name="Delta-Filter" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="4.0.0beta2">mummer4</requirement>
+        <requirement type="package" version="1.12.14">sh</requirement>
```
Comment (bgruening) [inline tools/mummer4/delta-filter.xml]: @nickeener why are all the other dependency needed? You just need mummer4 isn't it? And what is `sh`?

File: tools/mummer4/mummer.xml [hunk 4]
```diff
@@ -0,0 +1,176 @@
+<tool id="mummer" name="Mummer" version="0.1.0">
```
Comment (bgruening) [inline tools/mummer4/mummer.xml]: I would change all tool versions to the mummer version.

File: tools/mummer4/mummer.xml [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                        <param name="max_y" type="integer" argument="-y" value="100" label="Maximum Y-axis range" />
+                    </when>
+                    <when value="no" />
+                </conditional>
+            </when>
+            <when value="no" />
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="output" format="txt" from_work_dir="mummer.txt" label="mummer" />
```
Comment (bgruening) [inline tools/mummer4/mummer.xml]: I would remove all label from the outputs ... or use something like `label="${tool.name} on ${on_string}: mumplot"`

Comment (bgruening) [reply tools/mummer4/mummer.xml]: For all tools :)

File: tools/mummer4/nucmer.xml [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        <requirement type="package" version="1.12.14">sh</requirement>
+        <requirement type="package" version="4.4">sed</requirement>
+        <requirement type="package" version="4.2.1">gawk</requirement>
+        <requirement type="package" version="3.2.6a">fig2dev</requirement>
+        <requirement type="package" version="5.2.3">gnuplot</requirement>
+        <requirement type="package" version="3.2.6a">xfig</requirement>
+    </requirements>
+    <command detect_errors="exit_code">
+        <![CDATA[
+        #set $query=""+"".join(str($query_sequence).split(','))+""
```
Comment (bgruening) [inline tools/mummer4/nucmer.xml]: can you also use `"'+'"` here? Is the leading `+""` needed?

File: tools/mummer4/show-coords.xml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        </param>
+        <param name="switch" type="boolean" argument="-T" truevalue="-T" falsevalue="" label="Tab" help="Switch output to tab-delimited format" checked="true" />
+    </inputs>
+    <outputs>
+        <data name="output" format="txt" from_work_dir="show-coords.txt" label="show_coords" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="delta" ftype="txt" value="nucmer.txt" />
+            <output name="output" ftype="txt" compare="diff" value="show-coords.txt" />
```
Comment (bgruening) [inline tools/mummer4/show-coords.xml]: We could, if you like, attach the column names actually as metadata.
See here: https://github.com/galaxyproject/tools-iuc/blob/master/tools/deseq2/deseq2.xml#L170

Comment (bgruening) [reply tools/mummer4/show-coords.xml]: And it should probably be tabular isn't it?

**Comment by bgruening** [comment]
@nickeener I rebased against master for you, but a few tests are still failing. Sorry.

**Comment by nekrut** [comment]
@nickeener add help section as we discussed. Once done, @bgruening I think we are ready!

**Comment by nickeener** [comment]
Thanks for all the help @bgruening!

**Pull request merged**